### PR TITLE
File history

### PR DESCRIPTION
**Pull request by jdoliner**

### CODE REVIEW

**Comment by gabrielgrant** [comment]
Would it make sense to also re-introduce the `file.commit_modified` field that disappeared in 1.4?  #1457 Or is this a replacement? (meaning that issue should be closed)

**Comment by jdoliner** [comment]
I think this can effectively be a replacement, the `FileInfo.File.Commit` field effectively tells you the commit modified here, because you get back files at commits where they were modified. Note this isn't quite what we're returning now but it can be.

Another question I have on this is what happens if you delete a file. I.e. suppose I have a file, then delete it, then in a later commit readd it. If I go for the history of that file at `master` should I see the versions from before it was deleted? I think the answer is no because the pre-delete version is a different file. This is also nice from a performance perspective because it saves us having to always traverse back to the first ever commit to see if maybe a previous version of the file existed then.

**Comment by gabrielgrant** [comment]
summarizing offline convo:

right now if i get a list of files (without the new `--history` flag introduced by this PR), then `FileInfo.File.Commit` contains the commit passed into list-file. With the `--history` flag, I'll get a list of FileInfos, one for each commit at which the file was modified. In order to optimize this for the (common) case of just wanting to get the commit at which the file was most recently modified, @jdoliner proposed having the `--history` flag take an optional `number` parameter, defining how far to walk back in history. So calling list-file with `--history 1` would return a list of files in the head commit, each linked to the commit at which they were last modified

File: src/client/pfs/pfs.proto [hunk 1]
```diff
@@ -382,6 +382,19 @@ message ListFileRequest {
...
+  // semantics are:
+  // 0: return the files as they are at the commit in `file`. FileInfo.File
+  //    will equal File in this request.
+  // 1: return the files as they are in the last commit they were modified in.
+  //    (This will have the same hash as if you'd passed 0, but
+  //    FileInfo.File.Commit will be different.
+  // 2: Return the above and the files as they are in the next-last commit they
+  //    were modified in.
+  // 3: etc.
+  //-1: Return all historical versions.
```
[the remark this answers is not in the document]
Comment (jdoliner) [reply src/client/pfs/pfs.proto]: Good nitpick will change.

**Pull request merged**